### PR TITLE
Allow explictly setting head_dim instead of deducing it.

### PR DESCRIPTION
**Pull request by mbrenon**
This is a requirement for the upcoming OpenELM models.

Note that head_dim is also moved down to the attention config: this parameter is related to attention so it has no reason to live under the main model config. 

Please review carefully to make sure I'm not breaking anything here :)

BUG=https://b.corp.google.com/issues/352478939

### CODE REVIEW

File: ai_edge_torch/generative/layers/model_config.py [hunk 1]
```diff
@@ -55,9 +55,10 @@ class FeedForwardType(enum.Enum):
 
 @dataclass
 class AttentionConfig:
-  """Attention moduel's parameters."""
+  """Attention model's parameters."""
```
Comment (talumbau) [inline ai_edge_torch/generative/layers/model_config.py:58]: I think `module` is probably the best choice and follows PyTorch naming convention here. Both are better than the existing `moduel` though :-)

Comment (haozha111) [reply ai_edge_torch/generative/layers/model_config.py:58]: +1

File: ai_edge_torch/generative/layers/model_config.py [hunk 2]
```diff
@@ -55,9 +55,10 @@ class FeedForwardType(enum.Enum):
...
 
   num_heads: int
+  head_dim: int
```
Comment (haozha111) [inline ai_edge_torch/generative/layers/model_config.py:61]: i guess in most common cases (other than OpenELM) the `head_dim` could be automatically derived as the current library does, this will eliminate setting one more parameter when constructing the model.

Can we keep the current behavior (so that most models don't need a change and supply this additional parameter), but allow the parameter to be overwritten, or derived in another user supplied lamda?

File: ai_edge_torch/generative/layers/attention.py [hunk 3]
```diff
@@ -136,22 +136,24 @@ def __init__(
...
     self.config = config
     self.kv_cache = None
     self.batch_size = batch_size
+    qkv_shape = (config.num_heads + 2 * config.num_query_groups) * config.head_dim
+    output_shape = config.num_heads * config.head_dim
+    # Key, query, value projections for all heads.
+    self.qkv_projection = nn.Linear(dim, qkv_shape, bias=config.qkv_use_bias)
+    self.output_projection = nn.Linear(
+        output_shape, dim, bias=config.output_proj_use_bias
+    )
```
Comment (talumbau) [inline ai_edge_torch/generative/layers/attention.py:148]: Yes. This change is also needed for Gemma v2. Thanks @mbrenon! My plan will be to base my branch off this one.

File: ai_edge_torch/generative/layers/model_config.py (hunk 2 above)
Comment (haozha111) [reply ai_edge_torch/generative/layers/model_config.py:61]: this is kinda important design decision if we want to keep the `ModelConfig`api easy and simple to use, without exposing too many knobs to the user.

Comment (haozha111) [reply ai_edge_torch/generative/layers/model_config.py:61]: @talumbau @yichunk @hheydary any thoughts you could share here?

File: ai_edge_torch/generative/utilities/stable_diffusion_loader.py [hunk 4]
```diff
@@ -698,6 +698,7 @@ def load(
 
     attention_config = layers_config.AttentionConfig(
         num_heads=config.transformer_num_attention_heads,
+        head_dim=config.block_out_channels[0] // config.transformer_num_attention_heads,
```
Comment (talumbau) [inline ai_edge_torch/generative/utilities/stable_diffusion_loader.py:701]: This change will need @yichunk to review

File: ai_edge_torch/generative/layers/model_config.py (hunk 2 above)
Comment (talumbau) [reply ai_edge_torch/generative/layers/model_config.py:61]: GemmaV2 needs this exact change and I was about to make the changes to CausalSelfAttention that Maxime does here. I suspect more architectures will use this trick to squeeze down the parameter count a bit. So, I am +1 to being explicit here.

Comment (haozha111) [reply ai_edge_torch/generative/layers/model_config.py:61]: cool, sgtm. If this is going to be common, exposing it explicitly sounds like a good plan.

File: ai_edge_torch/generative/layers/attention.py (hunk 3 above)
Comment (mbrenon) [reply ai_edge_torch/generative/layers/attention.py:148]: Great!

File: ai_edge_torch/generative/utilities/stable_diffusion_loader.py (hunk 4 above)
Comment (mbrenon) [reply ai_edge_torch/generative/utilities/stable_diffusion_loader.py:701]: For context: I'm just using the way head_dim was previously defined [here](https://github.com/google-ai-edge/ai-edge-torch/blob/9fa04010af0aabfb1b97da1675202432155d138b/ai_edge_torch/generative/layers/model_config.py#L157-L158)

**Comment by talumbau** [comment]
I ran inference on the example decoder models with this branch and everything looked good. Also converted tiny_llama with no issues. Would be great to get an official thumbs up from someone who has worked with the Stable Diffusion model.

**Review by talumbau — APPROVED** [review]

**Comment by haozha111** [comment]
@yichunk could you help review the changes to SD code? thanks!

**Comment by talumbau** [comment]
Actually, I see we need another change for the T5 encoder/decoder models, similar to the other decoder-only models. Please add

```
 head_dim=64
```

To the `AttentionConfig` in get_model_config_t5() function in `t5.py`. Thanks @mbrenon!

**Comment by haozha111** [comment]
@yichunk could you help approve this PR? thanks!

File: ai_edge_torch/generative/examples/stable_diffusion/diffusion.py [hunk 5]
```diff
@@ -250,6 +250,7 @@ def __init__(self, config: unet_cfg.DiffusionModelConfig):
 
     attention_config = layers_cfg.AttentionConfig(
         num_heads=config.transformer_num_attention_heads,
+        head_dim=block_out_channels[0] // config.transformer_num_attention_heads,
```
Comment (yichunk) [inline ai_edge_torch/generative/examples/stable_diffusion/diffusion.py:253]: Attention dims are different for each block. Should construct attention config in the down_encoders loop below. Check `AttentionBlock2DConfig` and `CrossAttentionBlock2DConfig`.

Also, the attention configs in mid_block and up_decoders should be fixed as well.

File: ai_edge_torch/generative/utilities/stable_diffusion_loader.py (hunk 4 above)
Comment (yichunk) [reply ai_edge_torch/generative/utilities/stable_diffusion_loader.py:701]: Same as config defined in `diffusion.py`. The head dim will be different for each block.

**Comment by talumbau** [comment]
Closing in favor of #120
